src/dict.js

```diff
@@ -69,68 +99,63 @@ Sk.builtin.dict.tp$call = function(args, kw) {
 Sk.abstr.setUpInheritance("dict", Sk.builtin.dict, Sk.builtin.object);
 Sk.abstr.markUnhashable(Sk.builtin.dict);
 
-var kf = Sk.builtin.hash;
-
-Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
-
-    // Fast path: We spend a *lot* of time looking up strings
-    // in dictionaries. (Every attribute access, for starters.)
-    if (key.ob$type === Sk.builtin.str) {
-        for (i = 0; i < bucket.items.length; i++) {
-            item = bucket.items[i];
-            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
-                return item;
-            }
-        }
-        return null;
+function kf(key) {
+    // str => jsstr().replace(/^[0-9!#_]/, "!$&") avoids conflicts
+    // other => hash.v value from builtin.hash (javascript number)
+    let key_hash = key.$savedKeyHash_;
+    if (key_hash !== undefined) {
+        return key_hash;
+    } else if (key.ob$type === Sk.builtin.str) {
+        key_hash = key.$jsstr().replace(/^[0-9!#_]/, "!$&");
```

Comment:
You may also want to experiment will hoisting this regex out and pre-compiling it; might not matter, but I remember it making a performance difference in the parser.